powermon/protocols/pi18.py

```diff
@@ -112,9 +744,13 @@ def check_crc(self, response: str, command_definition: CommandDefinition = None)
             else:
                 log.info("PI18 response check_crc doesnt match got (%x, %x), calc (%x, %x)", crc_high, crc_low, response[-3], response[-2])
                 return False
+        elif response.startswith(b"^1") or response.startswith(b"^0"):
+           log.info("PI18 response doesnt start with ^D - but ACK or NACK ")
+           return True
         else:
             log.info("PI18 response doesnt start with ^D - check_crc fails")
             return False
```

Comment:
raise InvalidCRC

Reply:
But here is the problem.
I don't get b'^D1\x0b\xc2\r'  but a 
b'^1\x0b\xc2\r' for ACK and 
b'^0\x0b\xc2\r' for NACK

if response.startswith(b"^D") is not true
```
if response.startswith(b"^D"):
            # get response CRC
            data_to_check = response[:-3]
            crc_high, crc_low = crc(data_to_check)
            # print(crc_high, crc_low)
            # print(response[-3], response[-2])
            if (crc_high, crc_low) == (response[-3], response[-2]):
                return True
            else:
                log.info("PI18 response check_crc doesnt match got (%x, %x), calc (%x, %x)", crc_high, crc_low, response[-3], response[-2])
                return False
        elif response.startswith(b"^1") or response.startswith(b"^0"):
           log.info("PI18 response doesnt start with ^D - but ACK or NACK ")
           return True
```
this works also
`if response.startswith(b"^D") or response.startswith(b"^1") or response.startswith(b"^0"):`


```
2024-05-13 01:59:18,799:INFO:pi18:get_full_command@773: Using protocol: b'PI18' with 21 commands
2024-05-13 01:59:18,800:DEBUG:abstractprotocol:get_command_definition@108: Matched: BUCD500,530 to: BUCD value: 500,530
2024-05-13 01:59:18,800:DEBUG:helpers:crc_pi30@25: Calculating CRC for b'^S014BUCD500,530'
2024-05-13 01:59:18,801:DEBUG:helpers:crc_pi30@75: Generated CRC 0xf0 0x57 0xf057
2024-05-13 01:59:18,802:DEBUG:pi18:get_full_command@798: full command: b'^S014BUCD500,530\xf0W\r'
2024-05-13 01:59:18,802:DEBUG:usbport:send_and_receive@108: length of to_send: 19
2024-05-13 01:59:18,803:DEBUG:usbport:send_and_receive@118: multiple chunk send
2024-05-13 01:59:18,803:DEBUG:usbport:send_and_receive@125: sending chunk: b'^S014BUC'
2024-05-13 01:59:18,856:DEBUG:usbport:send_and_receive@125: sending chunk: b'D500,530'
2024-05-13 01:59:18,912:DEBUG:usbport:send_and_receive@125: sending chunk: b'\xf0W\r\x00\x00\x00\x00\x00'
2024-05-13 01:59:19,368:DEBUG:usbport:send_and_receive@143: usb response was: b'^1\x0b\xc2\r'
2024-05-13 01:59:19,368:DEBUG:command:build_result@153: build_result: for command with 'code: BUCD500,530, command_definition: CommandDefinition: self.code='BUCD', self.description='Set Battery Stop dis,charging when Grid is available', self.result_type=<ResultType.ACK: 2>, reading_definition count: 1, self.command_code=None, self.command_type=<CommandType.PI18_SETTER: 'pi18_setter'>'
2024-05-13 01:59:19,368:DEBUG:abstractprotocol:check_valid@155: check valid for b'^1\x0b\xc2\r', definition: CommandDefinition: self.code='BUCD', self.description='Set Battery Stop dis,charging when Grid is available', self.result_type=<ResultType.ACK: 2>, reading_definition count: 1, self.command_code=None, self.command_type=<CommandType.PI18_SETTER: 'pi18_setter'>
2024-05-13 01:59:19,368:DEBUG:pi18:check_crc@735: check crc for b'^1\x0b\xc2\r' in pi18
2024-05-13 01:59:19,368:DEBUG:helpers:crc_pi30@25: Calculating CRC for b'^1'
2024-05-13 01:59:19,368:DEBUG:helpers:crc_pi30@75: Generated CRC 0x0b 0xc2 0x0bc2
2024-05-13 01:59:19,368:DEBUG:pi18:trim_response@759: trim b'^1\x0b\xc2\r', definition: CommandDefinition: self.code='BUCD', self.description='Set Battery Stop dis,charging when Grid is available', self.result_type=<ResultType.ACK: 2>, reading_definition count: 1, self.command_code=None, self.command_type=<CommandType.PI18_SETTER: 'pi18_setter'>
2024-05-13 01:59:19,368:DEBUG:abstractprotocol:split_response@175: splitting b'^1', result_type ResultType.ACK
2024-05-13 01:59:19,368:DEBUG:abstractprotocol:split_response@230: responses: 'b'^1''
2024-05-13 01:59:19,369:INFO:result:decode_responses@104: result.response passed to decode: b'^1', result_type ResultType.ACK
2024-05-13 01:59:19,369:DEBUG:command_definition:get_reading_definition@116: looking for reading definition with: None, result_type is: ResultType.ACK
2024-05-13 01:59:19,369:DEBUG:command_definition:get_reading_definition@134: found reading definition: self.index=0, self.description='Set Battery Stop dis,charging when Grid is available', self.response_type=<ResponseType.ACK: 'ack'>, self.unit='', instance=<class 'powermon.commands.reading_definition.ReadingDefinitionACK'>
2024-05-13 01:59:19,369:DEBUG:result:decode_responses@161: got readings:
2024-05-13 01:59:19,369:DEBUG:result:__init__@57: Result: Result: self.command.command_definition.description='Set Battery Stop dis,charging when Grid is available': self.is_valid=True, self.error=False - self.error_messages=[], self.raw_response=b'^1\x0b\xc2\r',
2024-05-13 01:59:19,369:DEBUG:abstractport:run_command@92: after send_and_receive: Result: self.command.command_definition.description='Set Battery Stop dis,charging when Grid is available': self.is_valid=True, self.error=False - self.error_messages=[], self.raw_response=b'^1\x0b\xc2\r',
2024-05-13 01:59:19,369:INFO:device:run@196: Got result: Result: self.command.command_definition.description='Set Battery Stop dis,charging when Grid is available': self.is_valid=True, self.error=False - self.error_messages=[], self.raw_response=b'^1\x0b\xc2\r',
2024-05-13 01:59:19,369:DEBUG:device:run@212: Using Output: outputs.Screen: outputs the results to the screen as per the formatter supplied
2024-05-13 01:59:19,369:INFO:screen:process@19: Using output sender: screen
2024-05-13 01:59:19,369:DEBUG:screen:process@20: formatter: outputs the results to standard out in a table (optionally formatted with line art boxes), result: Result: self.command.command_definition.description='Set Battery Stop dis,charging when Grid is available': self.is_valid=True, self.error=False - self.error_messages=[], self.raw_response=b'^1\x0b\xc2\r', , mqtt_broker: MqttBroker DISABLED, device_info: <powermon.device.DeviceInfo object at 0x7f7d7a2ff200>
2024-05-13 01:59:19,369:INFO:table:format@29: Using output formatter: table
2024-05-13 01:59:19,370:DEBUG:table:format@45: displayData: Reading: self.data_name='Error', self.data_value='No readings in result', self.data_unit='', self.is_valid=True
----------------------------------------------------------------------------
Command: BUCD500,530 - Set Battery Stop dis,charging when Grid is available
----------------------------------------------------------------------------
Parameter  Value                  Unit
error      No readings in result
```

Reply:
the comment is related to `return False`
the failure approach for check_crc and check_valid is to raise an exception (weird I know, but that seems the best approach after trying multiple)

Reply:
something like
```python
if response == b'^1\x0b\xc2\r':
    log.debug('ACK response')
    return True
if response == b'^0\x0b\xc2\r'
    log.debug('NACK response')
    return True
if response.startswith(b"^D"):
    # get response CRC
    data_to_check = response[:-3]
    crc_high, crc_low = crc(data_to_check)
    # print(crc_high, crc_low)
    # print(response[-3], response[-2])
    if (crc_high, crc_low) == (response[-3], response[-2]):
        return True
    else:
        log.info("PI18 response check_crc doesnt match got (%x, %x), calc (%x, %x)", crc_high, crc_low, response[-3], response[-2])
        raise InvalidCRC(f"response has invalid CRC - got '\\x{response[-3]:02x}\\x{response[-2]:02x}', calculated '\\x{calc_crc_high:02x}\\x{calc_crc_low:02x}'", )
else:
    log.info("PI18 response doesnt start with ^D - check_crc fails")
    raise InvalidCRC("PI18 response doesnt start with ^D - check_crc fails")
```